---
title: "RNA-Seq_differential_expression_normalise_by_chromo"
output: html_document
date: "2025-09-30"
---

```{r,packages}


suppressPackageStartupMessages({
library("EnhancedVolcano")
library("limma")
library("biomaRt")
library("annotables")
library("tximport")
library("apeglm")
library("eulerr")
library("DESeq2")
library("HGNChelper")
library("tictoc")
library("DESeq2")
library("kableExtra")
library("beeswarm")
library("missMethyl")
library("gridExtra")
library("png")
library("metafor")
library("ggplot2")
library("purrr")
library("metafor")
library("dplyr")
library("readxl")
library("ggplot2")
library("tidyverse")
library("magrittr")
library("readr")
library("eulerr")
library("RColorBrewer")
library("e1071")
library("plotly") 
library("pheatmap")

})

CORES=16

```

# Read the sample manifest and alignment summary


```{r fig.width=11, fig.height=9, message=FALSE, warning=FALSE, echo=TRUE}

samples<-readRDS("/mnt/vol1/Projects_assist/counts_data.rds")

pheno<-readRDS("/mnt/vol1/Projects_assist/colData.rds")
pheno$ex_code <- rownames(pheno)

stopifnot(all(colnames(samples) %in% pheno$ex_code))
pheno <- pheno[match(colnames(samples), pheno$ex_code), ]
stopifnot(identical(colnames(samples), pheno$ex_code))

rownames(pheno) <- pheno$ex_code

pheno$Seq_Batch <- relevel(pheno$Seq_Batch, ref = "1")
pheno$Sex     <- relevel(pheno$Sex, ref = "Female")
pheno$age_bin <- relevel(pheno$age_bin, ref = "a18_29")

```

# Making DESeqDataSetFromMatrix


```{r fig.width=11, fig.height=9, message=FALSE, warning=FALSE, echo=TRUE}

dds_org <- DESeqDataSetFromMatrix(
  countData = samples,
  colData   = pheno,
  design    = ~ Seq_Batch + age_bin + Sex + age_bin:Sex
)

```


# Filtering for low count and normalisation of the counts

```{r fig.width=11, fig.height=9, message=FALSE, warning=FALSE, echo=TRUE}

keep <- rowSums(counts(dds_org)) >= 10
dds  <- dds_org[keep, ]

dds   <- estimateSizeFactors(dds)
norm_counts <- counts(dds, normalized = TRUE)

head(norm_counts)


```


# Unsupervised clustering analyses and DE analysis


```{r fig.width=11, fig.height=9, message=FALSE, warning=FALSE, echo=TRUE}

vsd <- vst(dds, blind = TRUE)

# Example correlation heatmap (annotation from pheno)
vsd_cor <- cor(assay(vsd))
pheatmap(
  vsd_cor,
  annotation_col = pheno[, c("Sex","age_bin","Seq_Batch")],
  annotation_row = pheno[, c("Sex","age_bin","Seq_Batch")]
)

# Fit model
dds <- DESeq(dds)

plotDispEsts(dds)


```


#  Model diagnostics


```{r fig.width=11, fig.height=9, message=FALSE, warning=FALSE, echo=TRUE}


# Sample-wise QC with VST 
vsd <- vst(dds, blind = TRUE)

# PCA
plotPCA(vsd, intgroup = c("Sex","age_bin","Seq_Batch"))
plotPCA(vsd, intgroup = c("Sex", "age_bin"))
plotPCA(vsd, intgroup = c("age_bin","Seq_Batch"))

```



```{r fig.width=11, fig.height=9, message=FALSE, warning=FALSE, echo=TRUE}

coefficients <- resultsNames(dds)
print(coefficients)

ref_age <- "a18_29"
sex_main <- "Sex_Male_vs_Female"

age_bins <- c("a30_39","a40_49","a50_59","a60_69","a70_79")

missing_itx <- setdiff(paste0("age_bin", age_bins, ".SexMale"), coefficients)
length(missing_itx)

# Shrinkage reference bin (a18_29)

results_list <- list()
results_list[[ref_age]] <- lfcShrink(dds, coef = sex_main, type = "apeglm")

# Other age bins use unshrunken list-contrast

for (age in age_bins) {
  int_term <- paste0("age_bin", age, ".SexMale")
  results_list[[age]] <- results(dds, contrast = list(c("Sex_Male_vs_Female", int_term)))
}

# 3) Filter significant DEGs (padj < 0.05 & |LFC| > 1)
results_sig_fc <- lapply(results_list, function(res) {
  res_df <- as.data.frame(res)
  subset(res_df, !is.na(padj) & padj < 0.05 & abs(log2FoldChange) > 1)
})



```

# Annotate DEGs with chromosome info

```{r fig.width=11, fig.height=9, message=FALSE, warning=FALSE, echo=TRUE}

summary_list <- list()

for (contrast_name in names(results_list)) {
  res    <- results_list[[contrast_name]] 
  res_df <- as.data.frame(res)
  res_df$ensembl_gene_id <- sub("\\..*$","", rownames(res_df))
  tested <- sum(!is.na(res_df$padj))

  sig_df <- results_sig_fc[[contrast_name]]
  if (is.null(sig_df)) sig_df <- res_df[0, , drop = FALSE] 
  sig_df$ensembl_gene_id <- sub("\\..*$","", rownames(sig_df))

  summary_list[[contrast_name]] <- data.frame(
    contrast    = contrast_name,
    tested_genes= tested,
    sig_total   = nrow(sig_df),
    sig_higher_in_male      = sum(sig_df$log2FoldChange > 0, na.rm = TRUE),
    sig_higher_in_female    = sum(sig_df$log2FoldChange < 0, na.rm = TRUE)
  )
}

summary_tbl <- do.call(rbind, summary_list)
print(summary_tbl)


```


# Annotate DEGs with chromosome info

```{r fig.width=11, fig.height=9, message=FALSE, warning=FALSE, echo=TRUE}

sig_ids <- unique(sub("\\..*$","", unlist(lapply(results_sig_fc, rownames), use.names = FALSE)))

map <- annotables::grch38[, c("ensgene","chr")]
map$ensgene <- sub("\\..*$","", map$ensgene)
chrom_map <- setNames(map$chr, map$ensgene)

# Fixed classes & palette
classes<- c("Autosome","X","Y","MT","Other")
pal<- setNames(RColorBrewer::brewer.pal(5, "Set2"), classes)

for (age in names(results_sig_fc)) {
  df <- results_sig_fc[[age]]
  if (is.null(df)) next
  if (nrow(df) == 0) next

  ids <- sub("\\..*$","", rownames(df))
  chr <- chrom_map[ids]
  cls <- ifelse(chr %in% as.character(1:22), "Autosome",
         ifelse(chr %in% c("X","Y"), chr,
         ifelse(chr %in% c("MT","M"), "MT", "Other")))

  pie <- as.data.frame(prop.table(table(factor(cls, levels = classes))) * 100)
  names(pie) <- c("chrom","prop")

   g <- ggplot(pie, aes("", prop, fill = chrom)) +
    geom_bar(stat = "identity", width = 1, color = "white") +
    coord_polar("y") +
    geom_text(aes(label = sprintf("%.1f%%", prop)),
              position = position_stack(0.5), size = 3) +
    scale_fill_manual(values = pal, drop = FALSE) +
    theme_void() +
    labs(title = paste("Chromosome distribution â€”", age), fill = "Chromosome")
     print(g)

}

```



# Annotate DEGs with chromosome info

```{r fig.width=11, fig.height=9, message=FALSE, warning=FALSE, echo=TRUE}


```


# Annotate DEGs with chromosome info

```{r fig.width=11, fig.height=9, message=FALSE, warning=FALSE, echo=TRUE}


```


